File: client/types/payload/request/register.request.ts
import { z } from "zod/v3";

export const RegisterRequest = z
  .object({
    username: z
      .string({ required_error: "Username cannot be blank" })
      .min(1, { message: "Username cannot be blank" })
      .min(5, {
        message: "Username is too short. Please enter at least 5 characters",
      })
      .max(50, {
        message:
          "Username is too long. Please enter no more than 50 characters",
      }),

    email: z
      .string({ required_error: "Email cannot be blank" })
      .min(1, { message: "Email cannot be blank" })
      .email({ message: "Please provide a valid email address" })
      .min(5, {
        message: "Email is too short. Please enter at least 5 characters",
      })
      .max(100, {
        message: "Email is too long. Please enter no more than 100 characters",
      }),

    password: z
      .string({ required_error: "Password cannot be blank" })
      .min(1, { message: "Password cannot be blank" })
      .min(6, {
        message: "Password is too short. Please enter at least 6 characters",
      })
      .max(100, {
        message:
          "Password is too long. Please enter no more than 100 characters",
      }),
    confirmPassword: z.string(),
  })
  .refine((data) => data.password === data.confirmPassword, {
    message: "Password does not match",
    path: ["confirmPassword"],
  })
  .transform(({ confirmPassword, ...data }) => data);

export type RegisterRequest = z.infer<typeof RegisterRequest>;

File: backend/src/main/java/com/lerneon/backend/models/payload/request/RegisterRequest.java
package com.lerneon.backend.models.payload.request;

import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class RegisterRequest {
    @NotBlank(message = "Username cannot be blank")
    @Size(min = 5, message = "Username is too short. Please enter at least 5 characters")
    @Size(max = 50, message = "Username is too long. Please enter no more than 50 characters")
    private String username;

    @NotBlank(message = "Email cannot be blank")
    @Email(message = "Please provide a valid email address")
    @Size(min = 5, message = "Email is too short. Please enter at least 5 characters")
    @Size(max = 100, message = "Email is too long. Please enter no more than 100 characters")
    private String email;

    @NotBlank(message = "Password cannot be blank")
    @Size(min = 6, message = "Password is too short. Please enter at least 6 characters")
    @Size(max = 100, message = "Password is too long. Please enter no more than 100 characters")
    private String password;
}
